File: 500/524 - Longest Word in Dictionary through Deleting.java
class Solution {
    public String findLongestWord(String s, List<String> d) {
        String ret = "";
        char[] array = s.toCharArray();
        int i;
        
        for (String word : d) {
            i = 0;
            
            for (char c : array) {
                if (i < word.length() && c == word.charAt(i))
                    ++i;
                
                if (i == word.length())
                    break;
            }
            
            if (i == word.length() && (word.length() > ret.length() || (word.length() == ret.length() && word.compareTo(ret) < 0)))
                ret = word;
        }
        
        return ret;
    }
}

File: csharp/524.longest-word-in-dictionary-through-deleting.cs
/*
 * @lc app=leetcode id=524 lang=csharp
 *
 * [524] Longest Word in Dictionary through Deleting
 */

using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {
    private static bool isSubsequence(string p, string s) {
        int m = p.Length, n = s.Length, i = 0, j = 0;
        if (m > n) {
            return false;
        }
        while (i < m && j < n) {
            if (p[i] == s[j]) {
                ++i;
            }
            ++j;
        }
        return i == m;
    }

    public string FindLongestWord(string s, IList<string> d) {
        foreach (string p in d.OrderByDescending(p => p.Length).ThenBy(p => p)) {
            if (isSubsequence(p, s)) {
                return p;
            }
        }
        return "";
    }
}
// @lc code=end

